x-pack/plugins/ml/server/models/json_schema_service/json_schema_service.ts

```diff
@@ -0,0 +1,154 @@
+
+const supportedEndpoints = [
+  {
+    path: '/_ml/anomaly_detectors/{job_id}' as const,
+    method: 'put',
+  },
+  {
+    path: '/_ml/datafeeds/{datafeed_id}' as const,
+    method: 'put',
+  },
+];
+
+/**
+ *
+ */
+export class JsonSchemaService {
+  /**
+   * Dictionary with the schema components
+   * @private
+   */
+  private _schemaComponents: Record<string, PropertyDefinition> = {};
+
+  private _requiredComponents: Set<string> = new Set<string>();
+
+  /**
+   * Extracts properties definition
+   * @private
+   */
+  private extractProperties(propertyDef: PropertyDefinition): void {
+    for (const key in propertyDef) {
+      if (propertyDef.hasOwnProperty(key)) {
+        const propValue = propertyDef[key as keyof PropertyDefinition]!;
+
+        if (key === '$ref') {
+          const comp = (propValue as string).split('/');
+          const refKey = comp[comp.length - 1];
+
+          delete propertyDef.$ref;
+
+          // FIXME there is an issue with the maximum call stack size exceeded
+          if (!refKey.startsWith('Ml_Types_')) return;
+
+          const schemaComponent = this._schemaComponents[refKey];
+
+          this._requiredComponents.add(refKey);
+
+          for (const k in schemaComponent) {
+            if (schemaComponent.hasOwnProperty(k)) {
+              // @ts-ignore
+              propertyDef[k] = schemaComponent[k] as PropertyDefinition[typeof k];
+            }
+          }
+          this.extractProperties(propertyDef);
+        }
+
+        if (Array.isArray(propValue)) {
+          propValue.forEach((v) => {
+            if (typeof v === 'object') {
+              this.extractProperties(v);
+            }
+          });
+        } else if (typeof propValue === 'object') {
+          this.extractProperties(propValue as PropertyDefinition);
+        }
+      }
+    }
+  }
+
+  private applyOverrides(path: SupportedPath, schema: PropertyDefinition): PropertyDefinition {
+    const overrides = jsonSchemaOverrides[path];
+    return {
+      ...schema,
+      ...overrides,
+      properties: {
+        ...schema.properties,
+        ...overrides.properties,
+      },
+    };
+  }
+
+  /**
+   * Extracts resolved schema definition for requested path and method
+   * @param path
+   * @param method
+   */
+  public async extractSchema(path: SupportedPath, method: string, schema?: object) {
+    const fileContent =
+      schema ?? JSON.parse(Fs.readFileSync(Path.resolve(__dirname, 'openapi.json'), 'utf8'));
+
+    const definition = fileContent.paths[path][method];
+
+    if (!definition) {
+      throw new Error('Definition is not defined');
+    }
+
+    const bodySchema = definition.requestBody.content['application/json'].schema;
+
+    this._schemaComponents = fileContent.components.schemas;
+
+    this.extractProperties(bodySchema);
+
+    return this.applyOverrides(path, bodySchema);
+  }
+
+  /**
+   * Generates openapi filed, removing redundant content.
+   */
+  public async generateSchemaFile() {
```

Comment:
Comment has been updated in https://github.com/elastic/kibana/pull/146968/commits/02d5bc94f7b5971babacd9b30d714820defa56c4

Reply:
What is the plan for future updates to the schema doc? Should this function be exposed by an endpoint and how does `openapi_source.json` get created?
It is possible that the schema will need to be updated by a person from the docs or java backend teams, so it would be good if it were as easy as possible to update.